pkg/segment/query/segquery.go

```diff
@@ -857,50 +889,31 @@ func applyAggOpOnSegments(sortedQSRSlice []*QuerySegmentRequest, allSegFileResul
 
 		var sstMap map[string]*structs.SegStats
+
 		if canUseSSTForStats(searchType, isSegmentFullyEnclosed, segReq.aggs) {
 			sstMap, err = segread.ReadSegStats(segReq.segKey, segReq.qid)
 			if err != nil {
-				log.Errorf("qid=%d,  applyAggOpOnSegments : ReadSegStats: Failed to get segment level stats for segKey %+v! Error: %v", qid, segReq.segKey, err)
+				log.Errorf("qid=%d, applyAggOpOnSegments: Failed to read segStats for segKey %+v! computing segStats from raw records. Error: %v",
+					qid, segReq.segKey, err)
 				allSegFileResults.AddError(err)
-				continue
-			}
-			sstMap["*"] = &structs.SegStats{
-				Count: uint64(segReq.TotalRecords),
-			}
-			allSegFileResults.AddResultCount(uint64(segReq.TotalRecords))
-		} else {
-			// run through micro index check for block tracker & generate SSR
-			blocksToRawSearch, err := segReq.GetMicroIndexFilter()
-			if err != nil {
-				log.Errorf("qid=%d, failed to get blocks to raw search! Defaulting to searching all blocks. SegKey %+v", segReq.qid, segReq.segKey)
-				blocksToRawSearch = segReq.GetEntireFileMicroIndexFilter()
-			}
-			sTime := time.Now()
-			isQueryPersistent, err := querytracker.IsQueryPersistent([]string{segReq.tableName}, segReq.sNode)
-			if err != nil {
-				log.Errorf("qid=%d, applyAggOpOnSegments: Failed to check if query is persistent! Error: %v", qid, err)
-			}
-			var rawSearchSSR map[string]*structs.SegmentSearchRequest
-			if segReq.sType == structs.SEGMENT_STATS_SEARCH {
-				rawSearchSSR = ExtractSSRFromSearchNode(segReq.sNode, blocksToRawSearch, segReq.queryRange, segReq.indexInfo.GetQueryTables(), qs, segReq.qid, isQueryPersistent, segReq.pqid)
+				// If we can't read the segment stats, we should compute it from raw records
+				sstMap, err = computeSegStatsFromRawRecords(segReq, qs, allSegFileResults, qid, nodeRes)
 			} else {
-				rawSearchSSR = metadata.ExtractUnrotatedSSRFromSearchNode(segReq.sNode, segReq.queryRange, segReq.indexInfo.GetQueryTables(), blocksToRawSearch, qs, segReq.qid)
-			}
-			qs.UpdateExtractSSRTime(time.Since(sTime))
-
-			// rawSearchSSR should be of size 1 or 0
-			for _, req := range rawSearchSSR {
-				req.ConsistentCValLenMap = segReq.ConsistentCValLenMap
-				sstMap, err = search.RawComputeSegmentStats(req, segReq.parallelismPerFile, segReq.sNode, segReq.segKeyTsRange, segReq.aggs.MeasureOperations, allSegFileResults, qid, qs, nodeRes)
-				if err != nil {
-					log.Errorf("qid=%d,  applyAggOpOnSegments : ReadSegStats: Failed to get segment level stats for segKey %+v! Error: %v", qid, segReq.segKey, err)
-					allSegFileResults.AddError(err)
+				sstMap["*"] = &structs.SegStats{
+					Count: uint64(segReq.TotalRecords),
 				}
+				allSegFileResults.AddResultCount(uint64(segReq.TotalRecords))
 			}
+		} else {
+			sstMap, err = computeSegStatsFromRawRecords(segReq, qs, allSegFileResults, qid, nodeRes)
 		}
+		if err != nil {
+			allSegFileResults.AddError(err)
```

Comment:
should this AddError be associated with the place where specifically the error occurred. Right now it could get added twice: line # 898 and 910

Reply:
addressed in the latest commit.